### PR TITLE
GEODE-5190: Refactor Serializable into separate interfaces

### PR DESCRIPTION
**Pull request by pivotal-jbarrett**
- Added DataSerializable
- Added DataSerializableFixeId
- Added DataSerializablePrimitive
- Added DataSerializableInternal

- Added IDataSerializable
- Added IDataSerializableFixeId
- Added IDataSerializablePrimitive
- Added IDataSerializableInternal
- Renamed IGeodeSerializable to ISerializable
- Renamed IGeodeDelta to IDelta

Co-authored-by: Ivan Godwin <igodwin@pivotal.io>

### CODE REVIEW

File: clicache/integration-test/ThinClientDeltaTestN.cs [hunk 1]
```diff
@@ -832,50 +832,55 @@ void runExpirationWithDelta()
 
     //#region Tests
 
-    [Test]
+    //[Test]
```
Comment (moleske) [inline clicache/integration-test/ThinClientDeltaTestN.cs]: I think I rather have tests that aren't working properly isolated to other file and marked in `CMakeLists.txt` as `omitted` with this commented in rather than commented out (assuming this isn't a mistake and is suppose to be in).  Same for the other `//[Test]`

File: clicache/integration-test2/SerializationTests.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            var cacheFactory = new CacheFactory();
+            Cache = cacheFactory.Create();
+            Cache.InitializeDeclarativeCache(CacheXml.File.FullName);
+        }
+
+        public void Dispose()
+        {
+            try
+            {
+                if (null != Cache)
```
Comment (moleske) [inline clicache/integration-test2/SerializationTests.cs]: Probably unnecessary since if `Cache` is null you have bigger problems, but depends where you value failure notifications.  I generally prefer throwing the null pointer error since you initialize the cache before each test.

File: clicache/integration-test2/SerializationTests.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                {
+                    Cache.Close();
+                }
+            }
+            finally
+            {
+                try
+                {
+                    if (null != CacheXml)
```
Comment (moleske) [inline clicache/integration-test2/SerializationTests.cs]: Same dealio as above.

File: tests/cli/NewFwkLib/PdxTest/PdxTests.cs [hunk 4]
```diff
@@ -231,7 +231,7 @@ protected TVal createIpdxInstance()
             pif.WriteByteArray("m_byte253", pt.Byte253);
             pif.WriteByteArray("m_byte65535", pt.Byte65535);
             pif.WriteByteArray("m_byte65536", pt.Byte65536);
-            pif.WriteObject("m_pdxEnum", pt.PdxEnum);
+            //pif.WriteObject("m_pdxEnum", pt.PdxEnum);
```
Comment (moleske) [inline tests/cli/NewFwkLib/PdxTest/PdxTests.cs]: just delete?

File: clicache/integration-test/ThinClientDeltaTestN.cs [hunk 5]
```diff
@@ -832,50 +832,55 @@ void runExpirationWithDelta()
...
     }
 
-    [Test]
+    //hangs [Test]
+    public void PutWithDeltaADWithCloning()
+    {
+      runDeltaWithAppdomian(true);
+    }
+
+    //passed [Test]
```
Comment (moleske) [inline clicache/integration-test/ThinClientDeltaTestN.cs]: like this one that says passed but is commented out?

File: clicache/integration-test2/SerializationTests.cs (hunk 2 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test2/SerializationTests.cs]: I am hoping to avoid the test framework giving me extra exceptions to track down should the creation of cache fail and then a I get a null reference trying to clean it up.

File: clicache/integration-test2/SerializationTests.cs (hunk 3 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test2/SerializationTests.cs]: Same as above.

File: tests/cli/NewFwkLib/PdxTest/PdxTests.cs (hunk 4 above)
Comment (pivotal-jbarrett) [reply tests/cli/NewFwkLib/PdxTest/PdxTests.cs]: Nope, need to add that back in. Good eye!

File: clicache/integration-test/ThinClientDeltaTestN.cs (hunk 5 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test/ThinClientDeltaTestN.cs]: Grrrfff... Will fix.

File: clicache/integration-test/ThinClientDeltaTestN.cs (hunk 1 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test/ThinClientDeltaTestN.cs]: Good plan.

File: clicache/integration-test2/SerializationTests.cs (hunk 2 above)
Comment (moleske) [reply clicache/integration-test2/SerializationTests.cs]: Cool, was thinking it would tell you what line during the throw, but also if you use cache in every test and it was null the exception would be thrown before this dispose anyway.  Carry on

File: clicache/integration-test/ThinClientDeltaTestN.cs (hunk 5 above)
Comment (moleske) [reply clicache/integration-test/ThinClientDeltaTestN.cs]: There's a few of those in this file, will let you go through them

Comment (pivotal-jbarrett) [reply clicache/integration-test/ThinClientDeltaTestN.cs]: Odd, I thought I checked in the uncommented version. Will double check.

Comment (pivotal-jbarrett) [reply clicache/integration-test/ThinClientDeltaTestN.cs]: Doh! Still on my machine.

File: clicache/integration-test/CMakeLists.txt [hunk 6]
```diff
@@ -100,7 +100,7 @@ set_property(TEST ThinClientQueryTestsN PROPERTY LABELS FLAKY)
 
 # Label any tests that always fail here
 set_property(TEST OverflowTestsN PROPERTY LABELS OMITTED)
-set_property(TEST ThinClientDeltaTestN PROPERTY LABELS OMITTED)
+set_property(TEST ThinClientDeltaTestFailing PROPERTY LABELS OMITTED)
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test/CMakeLists.txt]: No, we are moving away from that. Good catch though.

File: clicache/src/impl/ManagedCacheableKey.cpp [hunk 7]
```diff
@@ -79,7 +79,7 @@ namespace apache
           Apache::Geode::Client::GeodeException::ThrowNative(ex);
         }
       }
-
+
```
Comment (mcmellawatt) [inline clicache/src/impl/ManagedCacheableKey.cpp]: Extra spaces

File: clicache/src/impl/PdxInstanceFactoryImpl.cpp [hunk 8]
```diff
@@ -61,175 +65,175 @@ namespace Apache
...
+          m_pdxType->AddVariableLengthTypeField(fieldName, "String", PdxFieldTypes::STRING);
           m_FieldVsValues->Add(fieldName, value);
           return this;
         }
 
         IPdxInstanceFactory^ PdxInstanceFactoryImpl::WriteObject(String^ fieldName, Object^ value)
         {
           isFieldAdded(fieldName);
-          m_pdxType->AddVariableLengthTypeField(fieldName, /*obj->GetType()->FullName*/"Object", PdxTypes::OBJECT);
+          m_pdxType->AddVariableLengthTypeField(fieldName, /*obj->GetType()->FullName*/"Object", PdxFieldTypes::OBJECT);
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxInstanceFactoryImpl.cpp]: Commented out code mid-method call

File: clicache/src/impl/PdxReaderWithTypeCollector.cpp [hunk 9]
```diff
@@ -281,26 +283,24 @@ namespace Apache
...
-          m_newPdxType->AddVariableLengthTypeField(fieldName, /*obj->GetType()->FullName*/"Collection", apache::geode::client::GeodeTypeIds::CacheableArrayList);
-          int position = m_pdxType->GetFieldPosition(fieldName, m_offsetsBuffer, m_offsetSize, m_serializedLength);
-          if (position != -1)
-          {
-            m_dataInput->AdvanceCursorPdx(position);
-            PdxLocalReader::ReadCollection(fieldName, obj);
-            m_dataInput->RewindCursorPdx(position);
-          }
-        }
+        //TODO serializable not valid PdxFieldType id
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxReaderWithTypeCollector.cpp]: Commented out code

File: clicache/src/impl/PdxReaderWithTypeCollector.hpp [hunk 10]
```diff
@@ -38,7 +38,7 @@ namespace Apache
         private:
           PdxType^      m_newPdxType;
         
-					void checkType( String^ fieldName, Byte typeId, String^ fieldType);
+					void checkType( String^ fieldName, PdxFieldTypes typeId, String^ fieldType);
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxReaderWithTypeCollector.hpp]: Formatting

File: clicache/src/impl/PdxReaderWithTypeCollector.hpp [hunk 11]
```diff
@@ -149,22 +149,18 @@ namespace Apache
...
-        
-        //TODO:
-        //virtual void WriteMap( String^ fieldName, System::Collections::IDictionary^ map );
 
         /// <summary>
         /// Read a collection from the data.
         /// </summary>
         /// <param name="fieldName">The name of a member field whose value to read.</param>
-        virtual void ReadCollection( String^ fieldName, System::Collections::IList^ collection)override;
+        //virtual void ReadCollection( String^ fieldName, System::Collections::IList^ collection)override;
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxReaderWithTypeCollector.hpp]: Commented out code

File: clicache/src/impl/PdxTypeRegistry.hpp [hunk 12]
```diff
@@ -39,64 +39,64 @@ namespace Apache
...
-
-            Int32 GetPDXIdForType(PdxType^ type, native::Pool*);
-
-            Int32 GetEnumValue(EnumInfo^ ei);
-
-            EnumInfo^ GetEnum(Int32 enumVal);
+          //test hook;
+          int testGetNumberOfPdxIds();
+
+          //test hook
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxTypeRegistry.hpp]: Out of scope maybe, but can we hide these test hooks?

File: clicache/src/impl/PdxTypeRegistry.hpp [hunk 13]
```diff
@@ -39,64 +39,64 @@ namespace Apache
...
+          Int32 GetEnumValue(EnumInfo^ ei);
+
+          EnumInfo^ GetEnum(Int32 enumVal);
 
         private:
-            Cache^ m_cache;
+          Cache^ m_cache;
 
-            IDictionary<Int32, PdxType^>^ typeIdToPdxType = gcnew Dictionary<Int32, PdxType^>();
+          Dictionary<Int32, PdxType^>^ typeIdToPdxType = gcnew Dictionary<Int32, PdxType^>();
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxTypeRegistry.hpp]: Did we mean to change this to Dictionary from IDictionary?  Usually using the interface is preferred in situations like this

File: clicache/src/impl/PdxWriterWithTypeCollector.cpp [hunk 14]
```diff
@@ -105,234 +105,232 @@ namespace Apache
...
-        //TODO:
-        //IPdxWriter^ PdxWriterWithTypeCollector::WriteMap( String^ fieldName, System::Collections::IDictionary^ map );
-
-        IPdxWriter^ PdxWriterWithTypeCollector::WriteCollection(String^ fieldName, System::Collections::IList^ obj)
-        {
-          m_pdxType->AddVariableLengthTypeField(fieldName, /*obj->GetType()->FullName*/"Collection", apache::geode::client::GeodeTypeIds::CacheableArrayList);
-          PdxLocalWriter::WriteCollection(fieldName, obj);
-          return this;
-        }
+        //TODO serializable not a PdxFieldTypeId
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxWriterWithTypeCollector.cpp]: Commented out code

File: clicache/src/impl/PdxWriterWithTypeCollector.cpp [hunk 15]
```diff
@@ -105,234 +105,232 @@ namespace Apache
...
-          m_pdxType->AddVariableLengthTypeField(fieldName, "stringASCIIHuge", PdxTypes::STRING);
+          m_pdxType->AddVariableLengthTypeField(fieldName, "stringASCIIHuge", PdxFieldTypes::STRING);
           PdxLocalWriter::WriteASCIIHuge(fieldName, value);
           return this;
         }
 
         IPdxWriter^ PdxWriterWithTypeCollector::WriteObject(String^ fieldName, Object^ obj)
         {
-          m_pdxType->AddVariableLengthTypeField(fieldName, /*obj->GetType()->FullName*/"Object", PdxTypes::OBJECT);
+          m_pdxType->AddVariableLengthTypeField(fieldName, /*obj->GetType()->FullName*/"Object", PdxFieldTypes::OBJECT);
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxWriterWithTypeCollector.cpp]: Commented out code mid-method

File: clicache/src/impl/PdxWriterWithTypeCollector.hpp [hunk 16]
```diff
@@ -188,25 +188,20 @@ namespace Apache
...
-          //virtual IPdxWriter^ WriteMap( String^ fieldName, System::Collections::IDictionary^ map );
-
+          
           /// <summary>
           /// Write an collection to the <c>IPdxWriter</c>.
           /// </summary>
           /// <param name="fieldName">The name of the field associated with the value.</param>
           /// <param name="collection">The collection to write.</param>
-          virtual IPdxWriter^ WriteCollection( String^ fieldName, System::Collections::IList^ collection) override;
+          //virtual IPdxWriter^ WriteCollection( String^ fieldName, System::Collections::IList^ collection) override;
```
Comment (mcmellawatt) [inline clicache/src/impl/PdxWriterWithTypeCollector.hpp]: Commented out code

File: clicache/src/impl/SafeConvert.hpp [hunk 17]
```diff
@@ -52,77 +52,59 @@ namespace Apache
...
         if (serializableObject == nullptr) return nullptr;
 
-        if (auto mg_obj = std::dynamic_pointer_cast<native::ManagedCacheableKeyGeneric>(serializableObject))
+        if (auto managedDataSerializable = std::dynamic_pointer_cast<native::ManagedCacheableKeyGeneric>(serializableObject))
         {
-          return mg_obj->ptr();
+          return managedDataSerializable->ptr();
         }
-        if (auto mg_obj_delta = std::dynamic_pointer_cast<native::ManagedCacheableDeltaGeneric>(serializableObject))
+			  else if (auto managedCacheableDeltaGeneric = std::dynamic_pointer_cast<native::ManagedCacheableDeltaGeneric>(serializableObject))
```
Comment (mcmellawatt) [inline clicache/src/impl/SafeConvert.hpp]: White space issues

File: cppcache/include/geode/Serializable.hpp [hunk 18]
```diff
@@ -123,19 +78,13 @@ class APACHE_GEODE_EXPORT Serializable {
 
   template <class _T>
   inline static std::shared_ptr<Serializable> create(
-      const std::shared_ptr<_T>& value) {
+      const std::shared_ptr<_T> &value) {
```
Comment (mcmellawatt) [inline cppcache/include/geode/Serializable.hpp]: Reference formatting

File: cppcache/include/geode/internal/DataSerializableInternal.hpp [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+class DataOutput;
+class DataInput;
+
+namespace internal {
+
+class APACHE_GEODE_EXPORT DataSerializableInternal
+    : public virtual Serializable {
+ public:
+  ~DataSerializableInternal() override = default;
+  virtual void toData(DataOutput &dataOutput) const = 0;
```
Comment (mcmellawatt) [inline cppcache/include/geode/internal/DataSerializableInternal.hpp]: Ref formatting, this line and the one below it

File: cppcache/src/Serializable.cpp [hunk 20]
```diff
@@ -27,25 +28,23 @@ namespace apache {
...
-    return static_cast<int8_t>(GeodeTypeIdsImpl::CacheableUserData2);
-  } else {
-    return static_cast<int8_t>(GeodeTypeIdsImpl::CacheableUserData4);
-  }
-}
-
-int8_t Serializable::DSFID() const {
-  return static_cast<int8_t>(GeodeTypeIdsImpl::FixedIDDefault);
-}
+// int8_t Serializable::typeId() const {
```
Comment (mcmellawatt) [inline cppcache/src/Serializable.cpp]: Commented out code

File: cppcache/src/SerializationRegistry.hpp [hunk 21]
```diff
@@ -28,19 +28,22 @@
...
 #include <geode/Delta.hpp>
 #include <geode/PdxSerializable.hpp>
+#include <geode/DataSerializable.hpp>
 
 #include "util/concurrent/spinlock_mutex.hpp"
 #include "NonCopyable.hpp"
 #include "GeodeTypeIdsImpl.hpp"
 #include "MemberListForVersionStamp.hpp"
 #include "config.h"
+//#include "PdxHelper.hpp"
```
Comment (mcmellawatt) [inline cppcache/src/SerializationRegistry.hpp]: Commented out code

File: cppcache/src/TcrEndpoint.cpp [hunk 22]
```diff
@@ -1042,30 +1042,31 @@ GfErrType TcrEndpoint::sendRequestWithRetry(
...
-            closeConnection(conn);
-          }
-          break;
-        } else {
-          error = GF_NOTCON;
-          epFailure = true;
-          createNewConn = true;
-        }
       }
+      //} catch (...) {
```
Comment (mcmellawatt) [inline cppcache/src/TcrEndpoint.cpp]: Commented out code

File: tests/CMakeLists.txt [hunk 23]
```diff
@@ -16,7 +16,7 @@
 project(nativeclient.tests LANGUAGES NONE)
 
 add_subdirectory(cpp)
-add_subdirectory(javaobject)
+#add_subdirectory(javaobject)
```
Comment (mcmellawatt) [inline tests/CMakeLists.txt]: Commented out code

File: tests/cli/NewFwkLib/CacheServer.cs [hunk 24]
```diff
@@ -4620,7 +4620,7 @@ public void DoEntryOperationsForMU()
 
       TKey key;
       TVal value;
-      //IGeodeSerializable val;
+      //ISerializable val;
```
Comment (mcmellawatt) [inline tests/cli/NewFwkLib/CacheServer.cs]: Commented out code

File: tests/cli/NewFwkLib/FunctionExecution/FunctionExecution.cs [hunk 25]
```diff
@@ -667,7 +667,7 @@ public void DoAddDestroyNewKeysFunction()
       ResetKey("distinctKeys");
       Int32 numKeys = GetUIntValue("distinctKeys");
       int clientNum = Util.ClientNum;
-      //IGeodeSerializable[] filterObj = new IGeodeSerializable[numKeys];
+      //ISerializable[] filterObj = new ISerializable[numKeys];
```
Comment (mcmellawatt) [inline tests/cli/NewFwkLib/FunctionExecution/FunctionExecution.cs]: Commented out code in this file, looks like just global find/replace but could be worth fixing since this PR touched the code

**Comment by mcmellawatt** [comment]
Can we make sure that all our existing C++ and .NET examples still build and run as expected as part of this change?  I know we had a few serializer/serializable tests that are likely affected.

File: clicache/integration-test/DefaultCacheableN.cs [hunk 26]
```diff
@@ -36,8 +36,8 @@ public CustomSerializableObject()
     }
   }
 
-  // VJR: TODO: IGeodeSerializable should be replaced by IPdxSerializable when ready
-  class DefaultType : IGeodeSerializable
+  // VJR: TODO: IDataSerializable should be replaced by IPdxSerializable when ready
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test/DefaultCacheableN.cs]: Eh.... test code.

File: clicache/integration-test/ThinClientDeltaTestFailing.cs [hunk 27]
```diff
@@ -0,0 +1,657 @@
+/*
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test/ThinClientDeltaTestFailing.cs]: Yes. The tests that passed before the changes were pulled out into the other test case and then made to pass in this branch.

File: clicache/integration-test/SerializationTestsN.cs [hunk 28]
```diff
@@ -502,7 +491,7 @@ public UInt64 ObjectSize
       }
     }
 
-    public UInt32 ClassId
+    public Int32 ClassId
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test/SerializationTestsN.cs]: Consistency with the server. There is no explicit mention of negative class id on the server. Left the documented contract in place until we refactor the internal table that keeps track and used negatives for internal objects.

File: clicache/integration-test/ThinClientDeltaTest.cs [hunk 29]
```diff
@@ -118,9 +118,9 @@ public void ToDelta(DataOutput output)
 
     #endregion
 
-    #region IGeodeSerializable Members
+    #region IDataSerializable Members
 
-    public uint ClassId
+    public int ClassId
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test/ThinClientDeltaTest.cs]: Same answer

File: clicache/src/CacheableDate.cpp [hunk 30]
```diff
@@ -93,12 +92,7 @@ namespace Apache
 
       bool CacheableDate::Equals(ICacheableKey^ other)
       {
-        if (other == nullptr ||
-            other->ClassId != GeodeClassIds::CacheableDate) {
-          return false;
-        }
-        return m_dateTime.Equals(static_cast<CacheableDate^>(
-          other)->m_dateTime);
+        return Equals((Object^) other);
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/CacheableDate.cpp]: `==` == reference equality, `Equals` == .NET object equality

File: clicache/src/CacheableFileName.hpp [hunk 31]
```diff
@@ -60,55 +62,28 @@ namespace Apache
             gcnew CacheableFileName(value) : nullptr);
         }
 
-        // Region: IGeodeSerializable Members
+        // Region: ISerializable Members
 
-        /// <summary>
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/CacheableFileName.hpp]: Yeah, but this is implementation of interface. Docs should come from interface unless there is some deviation from the standard for the impl.

File: clicache/src/CacheableHashMap.hpp [hunk 32]
```diff
@@ -34,14 +33,15 @@ namespace Apache
   {
     namespace Client
     {
+			namespace native = apache::geode::client;
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/CacheableHashMap.hpp]: 👍

File: clicache/src/CacheableHashSet.hpp [hunk 33]
```diff
@@ -65,7 +67,7 @@ namespace Apache
             GC::KeepAlive(this);
           }
 
-          virtual void FromData(DataInput^ input) override
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/CacheableHashSet.hpp]: *&^%$#@! .NET distinguishes between interfaces and base classes. You can only override a base class.  :(

File: clicache/src/DataOutput.cpp [hunk 34]
```diff
@@ -479,73 +480,73 @@ namespace Apache
...
         {
           WriteByte(typeId);
           WriteInt64((Int64)obj);
           return;
         }
-        case apache::geode::client::GeodeTypeIds::CacheableDate:
+        case GeodeTypeIds::CacheableDate:
         {
           //CacheableDate^ cd = gcnew CacheableDate((DateTime)obj);
           //  TODO: igfser mapping between generic and non generic
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/DataOutput.cpp]: No in scope

File: clicache/src/DataOutput.hpp [hunk 35]
```diff
@@ -482,7 +482,31 @@ namespace Apache
...
+          }
+        }
+
+				static inline int8_t getDataSerializableFixedIdDsCode(int32_t fixedId) {
+          if (fixedId <= std::numeric_limits<int8_t>::max() &&
+              fixedId >= std::numeric_limits<int8_t>::min()) {
+            return static_cast<int8_t>(GeodeTypeIdsImpl::FixedIDByte);
+          } else if (fixedId <= std::numeric_limits<int16_t>::max() &&
+                     fixedId >= std::numeric_limits<int16_t>::min()) {
+            return static_cast<int8_t>(GeodeTypeIdsImpl::FixedIDShort);
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/DataOutput.hpp]: 👍 good catch!

File: clicache/src/impl/ManagedCacheableDelta.cpp [hunk 36]
```diff
@@ -100,60 +99,19 @@ namespace apache
...
-        // [0xc0000000, 0xe0000000) is for FixedIDShort
-        // and [0xe0000000, 0xffffffff] is for FixedIDInt
-        // Note: depends on fact that FixedIDByte is 1, FixedIDShort is 2
-        // and FixedIDInt is 3; if this changes then correct this accordingly
-        System::UInt32 classId = m_managedSerializableptr->ClassId;
-        if (classId >= 0x80000000) {
-          return (int8_t)((classId - 0x80000000) / 0x20000000);
-        }
-        return 0;
+        throw std::exception("unreachable");
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/impl/ManagedCacheableDelta.cpp]: No, it is literally unreachable code. The compiler just doesn't know that.

File: clicache/src/impl/PdxFieldType.hpp [hunk 37]
```diff
@@ -127,18 +127,18 @@ namespace Apache
           }
 
           virtual bool Equals(Object^ otherObj) override;
-          virtual Int32 GetHashCode() override;
+          
+					virtual Int32 GetHashCode() override;
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxFieldType.hpp]: 👍

File: clicache/integration-test/DefaultCacheableN.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test/DefaultCacheableN.cs]: Eh... test code and global search and replace.

File: clicache/integration-test2/CMakeLists.txt [hunk 38]
```diff
@@ -28,11 +30,12 @@ add_library( ${PROJECT_NAME} SHARED
     GeodeServerTests.cs
     CacheXml.cs
     CacheXmlTests.cs
-    RegionTest.cs
     cache.xml
     geode.properties
     xunit.runner.json
     packages.config
+    RegionTest.cs
+	SerializationTests.cs
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/integration-test2/CMakeLists.txt]: *&^%! Visual studio!

👍

File: clicache/src/CacheFactory.cpp [hunk 39]
```diff
@@ -66,14 +70,12 @@ namespace Apache
...
-          cache->TypeRegistry->RegisterTypeGeneric(
-            native::GeodeTypeIds::PdxType,
-            gcnew TypeFactoryMethodGeneric(Apache::Geode::Client::Internal::PdxType::CreateDeserializable),
-            nullptr);
+
 
           DistributedSystem::ManagedPostConnect(cache);
           DistributedSystem::registerCliCallback();
-          Serializable::RegisterPDXManagedCacheableKey(cache);
+          auto&& cacheImpl = CacheRegionHelper::getCacheImpl(nativeCache.get());
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/CacheFactory.cpp]: Why not?

File: clicache/src/impl/PdxInstanceFactoryImpl.cpp (hunk 8 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxInstanceFactoryImpl.cpp]: Global search and replace here buddy. Left the rest alone.

File: clicache/src/impl/PdxReaderWithTypeCollector.hpp (hunk 11 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxReaderWithTypeCollector.hpp]: 👍

File: clicache/src/impl/PdxTypeRegistry.hpp (hunk 12 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxTypeRegistry.hpp]: PdxTypeRegistry is internal

File: clicache/src/impl/PdxTypeRegistry.hpp (hunk 13 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxTypeRegistry.hpp]: Yes, the compiler made us do that. I can' recall why it suddenly made us but it did. Something about casting from object to interface with generics.

File: clicache/src/impl/PdxWriterWithTypeCollector.cpp (hunk 14 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxWriterWithTypeCollector.cpp]: 👍

File: clicache/src/impl/PdxWriterWithTypeCollector.hpp (hunk 16 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/PdxWriterWithTypeCollector.hpp]: 👍

File: clicache/src/impl/SafeConvert.hpp (hunk 17 above)
Comment (pivotal-jbarrett) [reply clicache/src/impl/SafeConvert.hpp]: VS is having fun with us.

File: cppcache/include/geode/Serializable.hpp (hunk 18 above)
Comment (pivotal-jbarrett) [reply cppcache/include/geode/Serializable.hpp]: 👍

File: cppcache/src/ClientHealthStats.hpp [hunk 40]
```diff
@@ -21,45 +21,28 @@
...
+
+#include "util/Log.hpp"
+#include "GeodeTypeIdsImpl.hpp"
 
 namespace apache {
 namespace geode {
 namespace client {
 
-class ClientHealthStats : public Serializable {
+class ClientHealthStats : public internal::DataSerializableFixedId_t<
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply cppcache/src/ClientHealthStats.hpp]: Says you

File: cppcache/src/TcrEndpoint.cpp (hunk 22 above)
Comment (pivotal-jbarrett) [reply cppcache/src/TcrEndpoint.cpp]: 👍

File: tests/CMakeLists.txt (hunk 23 above)
Comment (pivotal-jbarrett) [reply tests/CMakeLists.txt]: 👍

**Comment by pivotal-jbarrett** [comment]
@mcmellawatt All examples compile.

File: clicache/src/impl/ManagedCacheableDelta.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (mcmellawatt) [reply clicache/src/impl/ManagedCacheableDelta.cpp]: So it won't compile without this exception?  Seems odd to me, but I don't have the full context.

File: clicache/src/CacheFactory.cpp (hunk 39 above)
[the remark this answers is not in the document]
Comment (echobravopapa) [reply clicache/src/CacheFactory.cpp]: was asking for a friend... ;)

**Comment by pivotal-jbarrett** [comment]
@echobravopapa all the formatting issues are on C++/CLI code so there is nothing quick about it. :(

File: clicache/src/impl/ManagedCacheableDelta.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply clicache/src/impl/ManagedCacheableDelta.cpp]: It won't compile without a return statement of throw. Throw is more appropriate.
